src/main/java/com/google/cloud/tools/appengine/v2/internal/cloudsdk/CloudSdkProcessFactoryFactory.java

```diff
@@ -0,0 +1,88 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.appengine.v2.internal.cloudsdk;
+
+import com.google.cloud.tools.appengine.cloudsdk.internal.args.GcloudArgs;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Start and return processes for use with a CliProcessManager or for use with any other external
+ * process manager. The Input and Output streams are not handled or configured by this class. This
+ * class is configured by {@link CloudSdkAppEngineRequestFactory}.
+ */
+public class CloudSdkProcessFactoryFactory {
+
+  private final String metricsEnvironment;
+  private final String metricsEnvironmentVersion;
+  private final Path credentialFile;
+  private final CloudSdkV2 sdk;
+
+  /**
+   * Configure a new factory for ProcessFactory creation.
+   */
+  public CloudSdkProcessFactoryFactory(CloudSdkV2 sdk, String metricsEnvironment,
+      String metricsEnvironmentVersion,
+      Path credentialFile) {
+    this.sdk = sdk;
+    this.metricsEnvironment = metricsEnvironment;
+    this.metricsEnvironmentVersion = metricsEnvironmentVersion;
+    this.credentialFile = credentialFile;
+  }
+
+  @VisibleForTesting
+  protected Map<String, String> getEnvironment() {
+    Map<String, String> environment = Maps.newHashMap();
+    // see crendential-file-override in gcloud arguments
+    if (credentialFile != null) {
+      environment.put("CLOUDSDK_APP_USE_GSUTIL", "0");
+    }
+    if (metricsEnvironment != null) {
+      environment.put("CLOUDSDK_METRICS_ENVIRONMENT", metricsEnvironment);
+    }
+    if (metricsEnvironmentVersion != null) {
+      environment.put("CLOUDSDK_METRICS_ENVIRONMENT_VERSION", metricsEnvironmentVersion);
+    }
+    return environment;
+  }
+
+  @VisibleForTesting
+  protected List<String> getAppCommand(List<String> params) {
+    List<String> command = Lists.newArrayList();
+    command.add(sdk.getGCloudPath().toString());
+    command.add("app");
+    command.addAll(params);
+    command.add("--format=yaml");
+    command.add("--quiet");
+    if (credentialFile != null) {
+      command.addAll(GcloudArgs.get("credential-file-override", credentialFile.toFile()));
+    }
+    return command;
+  }
+
+  public <C> CloudSdkProcessFactory newAppProcessFactory(C configuration,
+      ConfigurationTranslator<C> configurationTranslator) {
+    return new CloudSdkProcessFactory(
+        getAppCommand(configurationTranslator.translate(configuration)), getEnvironment());
```

Comment:
It feels verbose to pass in 2 objects just to have one do the single thing it's intended with the other one. How about the translator receiving the config as constructor parameter and the `translate` method to be parameterless? In this case I would suggest a different naming though, maybe `ConfigurationWrapper`?